src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/CpsPackageReferenceProject.cs

```diff
@@ -202,68 +209,68 @@ private void IgnoreUnsupportProjectReference(PackageSpec project)
-                   .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
-                   .ToList();
-
-                // get the transitive packages, excluding any already contained in the installed packages
-                List<PackageReference> transitivePackages = packageSpec
-                   .TargetFrameworks
-                   .SelectMany(f => GetTransitivePackageReferencesForFramework(f.FrameworkName, _installedPackages, _transitivePackages, targets))
-                   .GroupBy(p => p.PackageIdentity)
-                   .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
-                   .ToList();
-                return new ProjectPackages(installedPackages, transitivePackages);
+                return new ProjectPackages(Array.Empty<PackageReference>(), Array.Empty<TransitivePackageReference>());
             }
-            else
+
+            IList<LockFileTarget> targetsList = null;
+            if (IsInstalledAndTransitiveComputationNeeded)
             {
-                return new ProjectPackages(Array.Empty<PackageReference>(), Array.Empty<PackageReference>());
+                // clear the transitive packages cache, since we don't know when a dependency has been removed
+                ClearCache();
+                targetsList = await GetTargetsListAsync(token);
             }
+
+            var frameworkSorter = new NuGetFrameworkSorter();
+
+            List<PackageReference> installedPackages = packageSpec
+                .TargetFrameworks
+                .SelectMany(f => GetPackageReferencesForFramework(
+                    f.Dependencies,
+                    f.FrameworkName,
+                    _installedPackages,
+                    targetsList))
+                .GroupBy(p => p.PackageIdentity)
+                .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
+                .ToList();
+
+            // get the transitive packages, excluding any already contained in the installed packages
+            List<PackageReference> transitivePackages = packageSpec
+                .TargetFrameworks
+                .SelectMany(f => GetTransitivePackageReferencesForFramework(
+                    f.FrameworkName,
+                    _installedPackages,
+                    _transitivePackages,
+                    targetsList))
+                .GroupBy(p => p.PackageIdentity)
+                .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
+                .ToList();
+
+            List<TransitivePackageReference> transitivePackagesWithOrigins = transitivePackages
+                .Select(transitivePr => Tuple.Create(transitivePr, GetTransitivePackageOrigin(
+                    transitivePr,
+                    installedPackages,
+                    targetsList,
+                    token)))
+                .Select(te => MergeTransitiveOrigin(te.Item1, te.Item2))
+                .ToList();
```

Comment:
We need to make sure that the code that computes the transitive origins can be easily disabled for the experiment. Please make sure that all computations for this feature are factored in such a way that can be disabled with a simple boolean check.

Reply:
Added A/B Experimentation code.